service/raft/raft_group0.cc

```diff
@@ -618,6 +795,618 @@ persistent_discovery::persistent_discovery(raft::server_address self, const peer
+        if (!remote_versions) {
+            upgrade_log.info("synchronize_schema: finished early.");
+            co_return false;
+        }
+        upgrade_log.info("synchronize_schema: collected remote schema versions.");
+
+        auto my_version = db.get_version();
+        upgrade_log.info("synchronize_schema: my version: {}", my_version);
+
+        auto matched = std::erase_if(*remote_versions, [my_version] (const auto& p) { return p.second == my_version; });
+        upgrade_log.info("synchronize_schema: schema mismatches: {}. {} nodes had a matching version.", *remote_versions, matched);
+
+        if (remote_versions->empty()) {
+            upgrade_log.info("synchronize_schema: finished.");
+            co_return true;
+        }
+
+        // Note: if we successfully merged schema from everyone in earlier iterations, but our schema versions
+        // are still not matching, that means our version is strictly more up-to-date than someone else's version.
+        // In that case we could switch to a push mode instead of pull mode (we push schema mutations to them);
+        // on the other hand this would further complicate the code and I assume that the regular schema synchronization
+        // mechanisms (gossiping schema digests and triggering pulls on the other side) should deal with this case,
+        // even though it may potentially take a bit longer than a pro-active approach. Furthermore, the other side
+        // is also executing `synchronize_schema` at some point, so having this problem should be extremely unlikely.
+        if (last_pull_successful) {
+            if ((++num_attempts_after_successful_pull) > 3) {
+                upgrade_log.error(
+                        "synchronize_schema: we managed to pull schema from every other node, but our schema versions"
+                        " are still different. The other side must have an outdated version and fail to pull it for some"
+                        " reason. If this message keeps showing up, the internal upgrade-to-raft procedure is stuck;"
+                        " try performing a rolling restart of your cluster."
+                        " If that doesn't fix the problem, the system may require manual fixing of schema tables.");
+            }
+        }
+
+        last_pull_successful = true;
+        co_await max_concurrent_for_each(*remote_versions, max_concurrency, [&] (const auto& p) -> future<> {
+            auto& [addr, _] = p;
+
+            try {
+                // We're using `do_merge_schema_from`, not `merge_schema_from`, because the latter coalesces all merge requests,
+                // including those coming from other sources than us (e.g. gossiper). Also, when we timeout, we want to abort the RPC
+                // call that we started, then on retry start another RPC call; we don't want to abort other users of `merge_schema_from`.
+                upgrade_log.info("synchronize_schema: `do_merge_schema_from({})`", addr);
```

Comment:
cc @gleb-cloudius 

Gleb mentioned something about RPC module hanging indefinitely if the other side doesn't have TCP stack.

Reply:
Note that the timeout only aborts the RPC part, if we manage to enter the schema merge part it won't react to the abort.

Reply:
https://github.com/scylladb/scylladb/pull/10835#discussion_r928680878

> Won't this cause something like "tcp closed connection" error?

> If there is no tcp stack to answer your SYN it will take minutes until you will get the error.

Reply:
I think such problems should be solved on the RPC layer, e.g. by using keep-alive or a failure detector. Client code should not have to handle this, since it may not want to time out like in this case, in which case there is no good choice for the timeout, it's is either too large or too small, which creates its own issues.

Reply:
> What's wrong with coalescing done by merge_schema_from()?

If we want to timeout our request, and the requests are coalesced, there is no way to timeout just us without timing out the others.
Well, we can of course just abandon the request, but that misses the point - because when we retry, we'll just block on our previous request which we just abandoned, so it's like we never abandoned the previous request at all.

Of course, there's the debatable part of whether we want to timeout at all. I don't have a strong opinion on this, if schema request blocks on TCP stack we probably won't be able to proceed anyway because a node is down.

OTOH, timing out allows us to proceed if we concurrently remove the down node - indeed, observe that the algorithm here re-fetches the peer list on each iteration, so if a node goes down and the algorithm gets stuck, the admin can remove the node using regular `removenode` - without entering `recovery` and all that annoying stuff - and the algorithm will unblock.

Reply:
Removing a node should fail the RPC due to connection being closed (possibly by keep-alive).

Reply:
Keep-alive is the option which "takes minutes", IIUC?
I would be fine with this. Let's see what @gleb-cloudius says.

Reply:
> cc @gleb-cloudius
> 
> Gleb mentioned something about RPC module hanging indefinitely if the other side doesn't have TCP stack.

Not indefinitely, but for minutes. And if it is not cancelable you cannot stop the process while rpc hangs.

Reply:
> Keep-alive is the option which "takes minutes", IIUC? I would be fine with this. Let's see what @gleb-cloudius says.

"Keep alive" is not related to connection establishment.

Reply:
So should we go with the cancellable option here @gleb-cloudius ?

Reply:
> So should we go with the cancellable option here @gleb-cloudius ?

What is your scenario? Github does not give me context any longer? If the scenario is that if a node died and you want to exclude it on the next iteration timeout is the only sane option.

Reply:
> "Keep alive" is not related to connection establishment.

So we don't timeout connection establishment right now?

Reply:
> > "Keep alive" is not related to connection establishment.
> 
> So we don't timeout connection establishment right now?

Connection establishment in TCP has a timeout. This timeout is not controlled by the application. It depends on various global TCP parameters. On my machine it takes about 2 minutes to timeout. Now it we have a code that waits for such RPC to complete during shutdown it means that shutdown may take up to 2 minutes. If shutdown does not wait for the RPC and aborts RPC client instead then this is not the issue. So it all depends on the context which I am missing here.